File: tests/fixtures/cross-impl-pysbol3/roundtrip.py
#!/usr/bin/env python3
"""Round-trip wrapper for the pySBOL3 cross-implementation conformance harness.

Reads an SBOL 3 Turtle file via pySBOL3 and writes the parsed document back
out in the requested output format. Output goes to stdout.

Invoked from inside the Docker container built by ``Dockerfile``. pySBOL3
may emit warnings to stderr during parsing; those are filtered by the
regenerate binary, and stdout carries only the serialized output.

Usage::

    python3 roundtrip.py <input.ttl> <format>

Where ``format`` is one of ``turtle``, ``rdfxml``, ``jsonld``, ``ntriples``.
"""

from __future__ import annotations

import sys
from pathlib import Path

import sbol3


# pySBOL3 expresses output formats as RDFlib serializer names. These are
# the canonical strings accepted by ``sbol3.Document.write_string``.
FORMAT_TO_RDFLIB = {
    "turtle": "turtle",
    "rdfxml": "xml",
    "jsonld": "json-ld",
    "ntriples": "nt",
}


def main(argv: list[str]) -> int:
    if len(argv) < 3:
        print("usage: roundtrip.py <input.ttl> <format>", file=sys.stderr)
        print("formats: turtle, rdfxml, jsonld, ntriples", file=sys.stderr)
        return 2

    input_path = Path(argv[1])
    requested_format = argv[2].lower()
    rdflib_format = FORMAT_TO_RDFLIB.get(requested_format)
    if rdflib_format is None:
        print(f"unknown format: {requested_format}", file=sys.stderr)
        return 2

    doc = sbol3.Document()
    doc.read(str(input_path))
    serialized = doc.write_string(file_format=rdflib_format)
    sys.stdout.write(serialized)
    if not serialized.endswith("\n"):
        sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: tests/fixtures/cross-impl/RoundTrip.java
// Round-trip wrapper used by the cross-implementation conformance
// harness. Reads an SBOL 3 Turtle file via libSBOLj3 and writes the
// parsed document back out in the requested output format. By default
// the output goes to stdout; passing a third argument writes to file.
//
// Invoked from inside the Docker container built by Dockerfile.
// libSBOLj3 emits a few INFO-level log lines on startup (Hibernate
// Validator, SLF4J). Those go to stderr and are filtered by the
// regenerate binary; stdout holds only the serialized output.

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;

import org.sbolstandard.core3.entity.SBOLDocument;
import org.sbolstandard.core3.io.SBOLFormat;
import org.sbolstandard.core3.io.SBOLIO;

public class RoundTrip {
    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("usage: RoundTrip <input.ttl> <format> [output]");
            System.err.println("formats: turtle, rdfxml, jsonld, ntriples");
            System.exit(2);
        }
        File input = new File(args[0]);
        SBOLFormat outputFormat = parseFormat(args[1]);

        OutputStream out = (args.length >= 3)
                ? new FileOutputStream(args[2])
                : System.out;

        // Source fixtures are always Turtle; the format arg controls the
        // output serialization only.
        SBOLDocument document = SBOLIO.read(input, SBOLFormat.TURTLE);
        SBOLIO.write(document, out, outputFormat);

        if (args.length >= 3) {
            out.close();
        }
    }

    private static SBOLFormat parseFormat(String name) {
        switch (name.toLowerCase()) {
            case "turtle":
                return SBOLFormat.TURTLE;
            case "rdfxml":
                return SBOLFormat.RDFXML;
            case "jsonld":
                return SBOLFormat.JSONLD;
            case "ntriples":
                return SBOLFormat.NTRIPLES;
            default:
                System.err.println("unknown format: " + name);
                System.exit(2);
                return null; // unreachable
        }
    }
}
